File: binding/src/libqrawfont.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qrawfont.html
pub const qrawfont = struct {

    /// New constructs a new QRawFont object.
    pub fn New() QtC.QRawFont {
        return qtc.QRawFont_new();
    }


    /// New2 constructs a new QRawFont object.
    pub fn New2(fileName: []const u8, pixelSize: f64) QtC.QRawFont {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QRawFont_new2(fileName_str, pixelSize);
    }


    /// New3 constructs a new QRawFont object.
    pub fn New3(fontData: []u8, pixelSize: f64) QtC.QRawFont {
        const fontData_str = qtc.libqt_string{
    .len = fontData.len,
    .data = fontData.ptr,
};
return qtc.QRawFont_new3(fontData_str, pixelSize);
    }


    /// New4 constructs a new QRawFont object.
    pub fn New4(other: ?*anyopaque) QtC.QRawFont {
        return qtc.QRawFont_new4(@ptrCast(other));
    }


    /// New5 constructs a new QRawFont object.
    pub fn New5(fileName: []const u8, pixelSize: f64, hintingPreference: i32) QtC.QRawFont {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QRawFont_new5(fileName_str, pixelSize, @intCast(hintingPreference));
    }


    /// New6 constructs a new QRawFont object.
    pub fn New6(fontData: []u8, pixelSize: f64, hintingPreference: i32) QtC.QRawFont {
        const fontData_str = qtc.libqt_string{
    .len = fontData.len,
    .data = fontData.ptr,
};
return qtc.QRawFont_new6(fontData_str, pixelSize, @intCast(hintingPreference));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QRawFont_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QRawFont_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QRawFont_IsValid(@ptrCast(self));
    }

    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QRawFont_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QRawFont_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn FamilyName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QRawFont_FamilyName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qrawfont.FamilyName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn StyleName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QRawFont_StyleName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qrawfont.StyleName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Style(self: ?*anyopaque, ) i32 {
        return qtc.QRawFont_Style(@ptrCast(self));
    }

    pub fn Weight(self: ?*anyopaque, ) i32 {
        return qtc.QRawFont_Weight(@ptrCast(self));
    }

    pub fn GlyphIndexesForString(self: ?*anyopaque, text: []const u8) []const u8 {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
return qtc.QRawFont_GlyphIndexesForString(@ptrCast(self), text_str);
    }

    pub fn AdvancesForGlyphIndexes(self: ?*anyopaque, glyphIndexes: []const u8) []const u8 {
        return qtc.QRawFont_AdvancesForGlyphIndexes(@ptrCast(self), @ptrCast(glyphIndexes));
    }

    pub fn AdvancesForGlyphIndexes2(self: ?*anyopaque, glyphIndexes: []const u8, layoutFlags: i32) []const u8 {
        return qtc.QRawFont_AdvancesForGlyphIndexes2(@ptrCast(self), @ptrCast(glyphIndexes), @intCast(layoutFlags));
    }

    pub fn GlyphIndexesForChars(self: ?*anyopaque, chars: ?*anyopaque, numChars: i32, glyphIndexes: *u32, numGlyphs: *i32) bool {
        return qtc.QRawFont_GlyphIndexesForChars(@ptrCast(self), @ptrCast(chars), numChars, @ptrCast(glyphIndexes), @ptrCast(numGlyphs));
    }

    pub fn AdvancesForGlyphIndexes3(self: ?*anyopaque, glyphIndexes: *const u32, advances: ?*anyopaque, numGlyphs: i32) bool {
        return qtc.QRawFont_AdvancesForGlyphIndexes3(@ptrCast(self), @ptrCast(glyphIndexes), @ptrCast(advances), numGlyphs);
    }

    pub fn AdvancesForGlyphIndexes4(self: ?*anyopaque, glyphIndexes: *const u32, advances: ?*anyopaque, numGlyphs: i32, layoutFlags: i32) bool {
        return qtc.QRawFont_AdvancesForGlyphIndexes4(@ptrCast(self), @ptrCast(glyphIndexes), @ptrCast(advances), numGlyphs, @intCast(layoutFlags));
    }

    pub fn AlphaMapForGlyph(self: ?*anyopaque, glyphIndex: u32) QtC.QImage {
        return qtc.QRawFont_AlphaMapForGlyph(@ptrCast(self), glyphIndex);
    }

    pub fn PathForGlyph(self: ?*anyopaque, glyphIndex: u32) QtC.QPainterPath {
        return qtc.QRawFont_PathForGlyph(@ptrCast(self), glyphIndex);
    }

    pub fn BoundingRect(self: ?*anyopaque, glyphIndex: u32) QtC.QRectF {
        return qtc.QRawFont_BoundingRect(@ptrCast(self), glyphIndex);
    }

    pub fn SetPixelSize(self: ?*anyopaque, pixelSize: f64) void {
        qtc.QRawFont_SetPixelSize(@ptrCast(self), pixelSize);
    }

    pub fn PixelSize(self: ?*anyopaque, ) f64 {
        return qtc.QRawFont_PixelSize(@ptrCast(self));
    }

    pub fn HintingPreference(self: ?*anyopaque, ) i32 {
        return qtc.QRawFont_HintingPreference(@ptrCast(self));
    }

    pub fn Ascent(self: ?*anyopaque, ) f64 {
        return qtc.QRawFont_Ascent(@ptrCast(self));
    }

    pub fn CapHeight(self: ?*anyopaque, ) f64 {
        return qtc.QRawFont_CapHeight(@ptrCast(self));
    }

    pub fn Descent(self: ?*anyopaque, ) f64 {
        return qtc.QRawFont_Descent(@ptrCast(self));
    }

    pub fn Leading(self: ?*anyopaque, ) f64 {
        return qtc.QRawFont_Leading(@ptrCast(self));
    }

    pub fn XHeight(self: ?*anyopaque, ) f64 {
        return qtc.QRawFont_XHeight(@ptrCast(self));
    }

    pub fn AverageCharWidth(self: ?*anyopaque, ) f64 {
        return qtc.QRawFont_AverageCharWidth(@ptrCast(self));
    }

    pub fn MaxCharWidth(self: ?*anyopaque, ) f64 {
        return qtc.QRawFont_MaxCharWidth(@ptrCast(self));
    }

    pub fn LineThickness(self: ?*anyopaque, ) f64 {
        return qtc.QRawFont_LineThickness(@ptrCast(self));
    }

    pub fn UnderlinePosition(self: ?*anyopaque, ) f64 {
        return qtc.QRawFont_UnderlinePosition(@ptrCast(self));
    }

    pub fn UnitsPerEm(self: ?*anyopaque, ) f64 {
        return qtc.QRawFont_UnitsPerEm(@ptrCast(self));
    }

    pub fn LoadFromFile(self: ?*anyopaque, fileName: []const u8, pixelSize: f64, hintingPreference: i32) void {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
qtc.QRawFont_LoadFromFile(@ptrCast(self), fileName_str, pixelSize, @intCast(hintingPreference));
    }

    pub fn LoadFromData(self: ?*anyopaque, fontData: []u8, pixelSize: f64, hintingPreference: i32) void {
        const fontData_str = qtc.libqt_string{
    .len = fontData.len,
    .data = fontData.ptr,
};
qtc.QRawFont_LoadFromData(@ptrCast(self), fontData_str, pixelSize, @intCast(hintingPreference));
    }

    pub fn SupportsCharacter(self: ?*anyopaque, ucs4: u32) bool {
        return qtc.QRawFont_SupportsCharacter(@ptrCast(self), ucs4);
    }

    pub fn SupportsCharacter2(self: ?*anyopaque, character: QtC.QChar) bool {
        return qtc.QRawFont_SupportsCharacter2(@ptrCast(self), @ptrCast(character));
    }

    pub fn SupportedWritingSystems(self: ?*anyopaque, ) []const u8 {
        return qtc.QRawFont_SupportedWritingSystems(@ptrCast(self));
    }

    pub fn FontTable(self: ?*anyopaque, tagName: []const u8, allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QRawFont_FontTable(@ptrCast(self), @ptrCast(tagName));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qrawfont.FontTable: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn FontTable2(self: ?*anyopaque, tag: QtC.QFont__Tag, allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QRawFont_FontTable2(@ptrCast(self), @ptrCast(tag));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qrawfont.FontTable2: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn FromFont(font: ?*anyopaque) QtC.QRawFont {
        return qtc.QRawFont_FromFont(@ptrCast(font));
    }

    pub fn AlphaMapForGlyph2(self: ?*anyopaque, glyphIndex: u32, antialiasingType: i32) QtC.QImage {
        return qtc.QRawFont_AlphaMapForGlyph2(@ptrCast(self), glyphIndex, @intCast(antialiasingType));
    }

    pub fn AlphaMapForGlyph3(self: ?*anyopaque, glyphIndex: u32, antialiasingType: i32, transform: ?*anyopaque) QtC.QImage {
        return qtc.QRawFont_AlphaMapForGlyph3(@ptrCast(self), glyphIndex, @intCast(antialiasingType), @ptrCast(transform));
    }

    pub fn FromFont2(font: ?*anyopaque, writingSystem: i32) QtC.QRawFont {
        return qtc.QRawFont_FromFont2(@ptrCast(font), @intCast(writingSystem));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QRawFont_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/rawfont.html#types
pub const enums = struct {
    pub const AntialiasingType = enum {
        pub const PixelAntialiasing: i32 = 0;
        pub const SubPixelAntialiasing: i32 = 1;
    };

    pub const LayoutFlag = enum {
        pub const SeparateAdvances: i32 = 0;
        pub const KernedAdvances: i32 = 1;
        pub const UseDesignMetrics: i32 = 2;
    };

};

File: binding/src/libqrawfont.cpp
#include <QByteArray>
#include <QChar>
#include <QFont>
#define WORKAROUND_INNER_CLASS_DEFINITION_QFont__Tag
#include <QImage>
#include <QPainterPath>
#include <QPointF>
#include <QRawFont>
#include <QRectF>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QTransform>
#include <qrawfont.h>
#include "libqrawfont.h"
#include "libqrawfont.hxx"

QRawFont* QRawFont_new() {
	 return new QRawFont();
}

QRawFont* QRawFont_new2(const libqt_string fileName, double pixelSize) {
	 return new QRawFont(QString::fromUtf8(fileName.data, fileName.len), pixelSize);
}

QRawFont* QRawFont_new3(const libqt_string fontData, double pixelSize) {
	 return new QRawFont(QByteArray(fontData.data, fontData.len), pixelSize);
}

QRawFont* QRawFont_new4(const QRawFont* other) {
	 return new QRawFont(*other);
}

QRawFont* QRawFont_new5(const libqt_string fileName, double pixelSize, int hintingPreference) {
	 return new QRawFont(QString::fromUtf8(fileName.data, fileName.len), pixelSize, static_cast<QFont::HintingPreference>(hintingPreference));
}

QRawFont* QRawFont_new6(const libqt_string fontData, double pixelSize, int hintingPreference) {
	 return new QRawFont(QByteArray(fontData.data, fontData.len), pixelSize, static_cast<QFont::HintingPreference>(hintingPreference));
}

void QRawFont_OperatorAssign(QRawFont* self, const QRawFont* other) {
	self->operator=(*other);
}

void QRawFont_Swap(QRawFont* self, QRawFont* other) {
	self->swap(*other);
}

bool QRawFont_IsValid(const QRawFont* self) {
	return self->isValid();
}

bool QRawFont_OperatorEqual(const QRawFont* self, const QRawFont* other) {
	return self->operator==(*other);
}

bool QRawFont_OperatorNotEqual(const QRawFont* self, const QRawFont* other) {
	return self->operator!=(*other);
}

libqt_string QRawFont_FamilyName(const QRawFont* self) {
	QString _ret = self->familyName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QRawFont_StyleName(const QRawFont* self) {
	QString _ret = self->styleName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QRawFont_Style(const QRawFont* self) {
	return self->style();
}

int QRawFont_Weight(const QRawFont* self) {
	return self->weight();
}

libqt_list QRawFont_GlyphIndexesForString(const QRawFont* self, const libqt_string text) {
	auto _ret = self->glyphIndexesForString(QString::fromUtf8(text.data, text.len));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_list QRawFont_AdvancesForGlyphIndexes(const QRawFont* self, const libqt_list glyphIndexes) {
	auto _ret = self->advancesForGlyphIndexes(QList<unsigned int>());
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_list QRawFont_AdvancesForGlyphIndexes2(const QRawFont* self, const libqt_list glyphIndexes, int layoutFlags) {
	auto _ret = self->advancesForGlyphIndexes(QList<unsigned int>(), static_cast<QFlags<QRawFont::LayoutFlag>>(layoutFlags));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

bool QRawFont_GlyphIndexesForChars(const QRawFont* self, const QChar* chars, int numChars, unsigned int* glyphIndexes, int* numGlyphs) {
	return self->glyphIndexesForChars(chars, numChars, glyphIndexes, numGlyphs);
}

bool QRawFont_AdvancesForGlyphIndexes3(const QRawFont* self, const unsigned int* glyphIndexes, QPointF* advances, int numGlyphs) {
	return self->advancesForGlyphIndexes(glyphIndexes, advances, numGlyphs);
}

bool QRawFont_AdvancesForGlyphIndexes4(const QRawFont* self, const unsigned int* glyphIndexes, QPointF* advances, int numGlyphs, int layoutFlags) {
	return self->advancesForGlyphIndexes(glyphIndexes, advances, numGlyphs, static_cast<QFlags<QRawFont::LayoutFlag>>(layoutFlags));
}

QImage* QRawFont_AlphaMapForGlyph(const QRawFont* self, unsigned int glyphIndex) {
	return new QImage(self->alphaMapForGlyph(glyphIndex));
}

QPainterPath* QRawFont_PathForGlyph(const QRawFont* self, unsigned int glyphIndex) {
	return new QPainterPath(self->pathForGlyph(glyphIndex));
}

QRectF* QRawFont_BoundingRect(const QRawFont* self, unsigned int glyphIndex) {
	return new QRectF(self->boundingRect(glyphIndex));
}

void QRawFont_SetPixelSize(QRawFont* self, double pixelSize) {
	self->setPixelSize(pixelSize);
}

double QRawFont_PixelSize(const QRawFont* self) {
	return self->pixelSize();
}

int QRawFont_HintingPreference(const QRawFont* self) {
	return self->hintingPreference();
}

double QRawFont_Ascent(const QRawFont* self) {
	return self->ascent();
}

double QRawFont_CapHeight(const QRawFont* self) {
	return self->capHeight();
}

double QRawFont_Descent(const QRawFont* self) {
	return self->descent();
}

double QRawFont_Leading(const QRawFont* self) {
	return self->leading();
}

double QRawFont_XHeight(const QRawFont* self) {
	return self->xHeight();
}

double QRawFont_AverageCharWidth(const QRawFont* self) {
	return self->averageCharWidth();
}

double QRawFont_MaxCharWidth(const QRawFont* self) {
	return self->maxCharWidth();
}

double QRawFont_LineThickness(const QRawFont* self) {
	return self->lineThickness();
}

double QRawFont_UnderlinePosition(const QRawFont* self) {
	return self->underlinePosition();
}

double QRawFont_UnitsPerEm(const QRawFont* self) {
	return self->unitsPerEm();
}

void QRawFont_LoadFromFile(QRawFont* self, const libqt_string fileName, double pixelSize, int hintingPreference) {
	self->loadFromFile(QString::fromUtf8(fileName.data, fileName.len), pixelSize, static_cast<QFont::HintingPreference>(hintingPreference));
}

void QRawFont_LoadFromData(QRawFont* self, const libqt_string fontData, double pixelSize, int hintingPreference) {
	self->loadFromData(QByteArray(fontData.data, fontData.len), pixelSize, static_cast<QFont::HintingPreference>(hintingPreference));
}

bool QRawFont_SupportsCharacter(const QRawFont* self, unsigned int ucs4) {
	return self->supportsCharacter(ucs4);
}

bool QRawFont_SupportsCharacter2(const QRawFont* self, QChar* character) {
	return self->supportsCharacter(*character);
}

libqt_list QRawFont_SupportedWritingSystems(const QRawFont* self) {
	auto _ret = self->supportedWritingSystems();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_string QRawFont_FontTable(const QRawFont* self, const char* tagName) {
	QByteArray _qb = self->fontTable(tagName);
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QRawFont_FontTable2(const QRawFont* self, QFont__Tag* tag) {
	QByteArray _qb = self->fontTable(*tag);
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QRawFont* QRawFont_FromFont(const QFont* font) {
	return new QRawFont(QRawFont::fromFont(*font));
}

QImage* QRawFont_AlphaMapForGlyph2(const QRawFont* self, unsigned int glyphIndex, int antialiasingType) {
	return new QImage(self->alphaMapForGlyph(glyphIndex, static_cast<QRawFont::AntialiasingType>(antialiasingType)));
}

QImage* QRawFont_AlphaMapForGlyph3(const QRawFont* self, unsigned int glyphIndex, int antialiasingType, const QTransform* transform) {
	return new QImage(self->alphaMapForGlyph(glyphIndex, static_cast<QRawFont::AntialiasingType>(antialiasingType), *transform));
}

QRawFont* QRawFont_FromFont2(const QFont* font, int writingSystem) {
	return new QRawFont(QRawFont::fromFont(*font, static_cast<QFontDatabase::WritingSystem>(writingSystem)));
}

void QRawFont_Delete(QRawFont* self) {
    delete self;
}

